### PR TITLE
"Endless" def support for ruby 3

### PR DESCRIPTION
**Pull request by AndrewRadev**
This is mostly here to have a separate branch for testing while I get things (hopefully) right.

Raised in issue: https://github.com/vim-ruby/vim-ruby/issues/432

### CODE REVIEW

File: indent/ruby.vim [hunk 1]
```diff
@@ -108,10 +111,26 @@ let s:end_middle_regex = '\<\%(ensure\|else\|\%(\%(^\|;\)\s*\)\@<=\<rescue:\@!\>
...
+  elseif expand('<cword>') == 'do'
+        \ && getline(".") =~ '^\s*\<\(while\|until\|for\):\@!\>'
+    return 1
+  elseif getline('.') =~ s:ruby_endless_def
+    return 1
+  elseif getline('.') =~ '\<def\s\+\k\+([^)]*$'
+    " Then it's a `def method(` with a possible `) =` later
+    call search('\<def\s\+\k\+\zs(', 'W', line('.'))
+    normal! %
+    return getline('.') =~ ')\s*='
```
Comment (AndrewRadev) [inline indent/ruby.vim:127]: I wish I knew if searching restores the cursor position after it's done evaluating the skip expression. I feel like it has to keep separate track over where the cursor is while searching, but the documentation doesn't seem to say anything.

**Review by cstyles — COMMENTED** [review]
Hey, thanks for taking a crack at fixing this! I haven't had a chance to play with it much yet but I did find one bug. The parentheses following the method name are only necessary if there are parameters but the new regex always treats the parentheses as necessary. So the next line after something like this...

```ruby
def foo = bar + 2
```

...will be indented even though it shouldn't.

**Comment by AndrewRadev** [comment]
Hm, I'd thought this wasn't allowed, but it definitely is -- I've adjusted the regex.

**Comment by AndrewRadev** [comment]
I'll merge this PR now to see if other people run into issues and keep #432 open for another week or two.

**Comment by marzdrel** [comment]
@AndrewRadev I've tested this for a bit and it seems to work pretty great, thanks!

**Comment by dkearns** [comment]
I was just having a look at updating the syntax file for these and noticed that the `=` is allowed on the following line when there are parentheses.

```
def foo()
  = 42
```

I wouldn't lose sleep over it but I know you like to be thorough. :)

**Comment by AndrewRadev** [comment]
I've added a note for it in the tests, but I'll skip this case for now. I feel like it'll take a while before the style guides come up with it as a recommendation 😅.

Interestingly, this doesn't even work with variable assignment. This gives me a syntax error:

``` ruby
foo
  = 1
```

**Pull request merged**